### ISSUE TITLE
Detect channel disconnects and user presence 

### ISSUE THREAD

**Issue by bufke** [jonashagstedt/swampdragon#70]
Is there anyway to know when a client disconnects? This would help enable presence detection.

**Comment by jonashagstedt** [comment]
If you want to do this in Python you can write a custom connection.
You can override `on_open` and `on_close` to do this.

This was briefly discussed here:
https://github.com/jonashagstedt/swampdragon/issues/68

If what you want to do is track online visitors this can be a bit complicated.

** The problem **:
The number of connections are per process, so if you were to depend on that you could never scale it over multiple processes.

If you use Redis to track the users what happens if the Redis connection goes down or if the `on_close` doesn't fire on the connection (not sure when this happens but it has happened during debugging).

** The solution **:
I've solved this in https://github.com/jonashagstedt/swampdragon-notifications/ and @basilkohler has a solution as well which I suspect is similar to mine.

In short:
Track the users using the `on_open` / `on_close` events on a custom connection (you have to write a custom connection), and make use of the heartbeat as well to clean up the users that are no longer connected.

If you look in SD notifications:
This is the connection:
https://github.com/jonashagstedt/swampdragon-notifications/blob/master/swampdragon_notifications/notification_connection.py
and here is the online user manager:
https://github.com/jonashagstedt/swampdragon-notifications/tree/master/swampdragon_notifications/online

Note that this makes use of swampdragon auth.
You can use a uuid or just the session id if you wan to allow anon users.

Hope this helps

**Comment by bufke** [comment]
Wow thanks for the thorough reply! Sounds like exactly what I want. I posted my playing around with chat [here](https://github.com/bufke/chat-experiment).

**Comment by bufke** [comment]
I started playing round with this to implement presence. I have notifications working. Could you comment on my approach? If I can think of any way to contribute back upstream I'd be happy to based on this work. If nothing else maybe a blog post explaining the process.

For this use case users see only a filtered set of presence data (friends)

1. Add a online status field to the user model. This adds interactions with the database but allows us to use self publishing models. Which is needed anyway in this use case to be aware of other users.
2. Extend the online user manager to set online status on change.

An issue I have is that user_manager is just a collection of functions. That makes it a bit harder to extend. Do you think this approach makes sense? If it does would it be useful to make user_manager a class that can be swapped out like the Connection can be?

**Comment by jonashagstedt** [comment]
If I understand you correctly you want to be able to access online users based on friends.

So not only see who is online but also filter based on a friendship setting.
Since Redis is so fast it might be worth keeping relations in Redis, however that's a side note.

If you are just doing presence you could write your own manager that filters out online users based on friends.

Eg.

    online_users  = get_online_users()  # from the user_manager
    friend_ids = user.get_friend_ids()
    friends_online = [u for u in online_users if u in friend_ids]

This should result in a list of friends who are also online.

I have to stress the point that things can go wrong along the way when marking a user as online so you can end up with a user being shown as online even though they have dropped offline (if the Redis server goes down or something else goes wrong along the way).
This is cleaned up by the heart beat but it means that a tiny tiny percentage of online users **might** actually be offline. It's very rare but I have to stress that it's a possibility.

I would say just write a new user manager that makes use of the existing functionality in the current `user_manager` but also adds your own functionality.

I would of course love to read a blog post on the topic!

**Comment by bufke** [comment]
Very helpful again, thanks. Cleaning things up on heartbeats makes sense. How can I create my own user_manager for this? To make sure I understand things I made my own Connection class and successfully got the user id from on_heartbeat, on_open, and on_close. From this I could make my own manager to detect presence. However it would be cleaner to just extend the redis user manager which already has 90% of the logic needed. 

redis_user_manager is a module and not a class. So I can't just extend it. swampdragon-notifications references it in a few places. Even if I copy the redis_user_manager code to use for my custom Connection class, the old module is still accessed in places like [here](https://github.com/jonashagstedt/swampdragon-notifications/blob/b60170a608ed3c5b8e798967bdf0b3279dde0df4/swampdragon_notifications/routers.py#L4). Trying to think of a clean solution here.

Should the user manager be a class instead for easier extending?

Alternatively I could unregister the OnlineUsersRouter (which doesn't seem supported right now?) and register my own. Realistically I need to extend OnlineUsersRouter to be more useful for me anyway so this isn't a big deal. Django admin has an unregister function so it would be consistent with that.

I could send some pull requests if any of these solutions makes sense.

**Comment by jonashagstedt** [comment]
Swampdragon-notifications is still quite new so there is room for improvement. I think we move this conversation to that repository instead as it's more related to that than core swampdragon.

https://github.com/jonashagstedt/swampdragon-notifications/issues/3
